### PR TITLE
refactor(datafiles): use 0-based kstp/kper indexing

### LINKED ISSUE #1338: hds.get_kstpkper() returns 0 based kstp,kper while hds.kstpkper returns one based.
hds.get_kstpkper() returns 0 based kstp,kper while hds.kstpkper returns one based. 
hds = fp.utils.binaryfile.HeadFile(r"my.hds")

Is there a good reason for this? If not I would think both should be 0 based like most other stuff in flopy?

### PR DESCRIPTION
**Pull request by wpbonelli**
Prompted by #1338

Plan originally proposed/implemented but abandoned (see discussion below):

* use 0-based indexing for both `kstpkper` attribute and `get_kstpkper()` method
* deprecate `get_kstpkper()` method (redundant)
* applies to
  * `LayerFile` and children
  * `BinaryLayerFile` and children
  * `CellBudgetFile`

### CODE REVIEW

**Review by langevin-usgs — COMMENTED** [review]
My recommendation on this would be to deprecate the `.kstpkper' member variable, since it doesn't behave the way we want anyway (1-based), and keep the `.get_kstpkper()` method instead.  Maybe we put a leading underscore in front of `._kstpkper`?  In the modflow6 repo we use the get_kstpkper() method pretty widely.

**Comment by wpbonelli** [comment]
Ok will leave `get_kstpkper()` in place. Looking at the base `LayerFile` the pattern looks like public properties &mdash; maybe just a change of `kstpkper` to 0-based for consistency and not deprecate or underscore/private? I don't think it's bad to have both, just thought to do opportunistic cleanup.

**Comment by wpbonelli** [comment]
maybe it would be safest just to clearly document that `kstpkper` is 1-based and leave it that way

**Comment by langevin-usgs** [comment]
> maybe it would be safest just to clearly document that `kstpkper` is 1-based and leave it that way

I think it's preferable to leave `.kstpkper` as one-based in case someone is using it now and relying on that behavior.  We might also consider deprecating it, as a way to continue removing remnant one-based information.  If we deprecate/remove it, those using it will have to switch to the zero-based `get_kstpkper()` method.

File: flopy/utils/formattedfile.py [hunk 1]
```diff
@@ -356,10 +356,9 @@ class FormattedHeadFile(FormattedLayerFile):
     >>> import flopy.utils.formattedfile as ff
     >>> hdobj = ff.FormattedHeadFile('model.fhd', precision='single')
     >>> hdobj.list_records()
-    >>> rec = hdobj.get_data(kstpkper=(1, 50))
+    >>> rec = hdobj.get_data(kstpkper=(0, 50))
```
Comment (jdhughes-usgs) [inline flopy/utils/formattedfile.py:359]: Although this is a made up example wouldn't it be consistent to change (1,50) to (0,49)

File: flopy/utils/binaryfile.py [hunk 2]
```diff
@@ -609,7 +611,7 @@ class HeadFile(BinaryLayerFile):
     >>> import flopy.utils.binaryfile as bf
     >>> hdobj = bf.HeadFile('model.hds', precision='single')
     >>> hdobj.list_records()
-    >>> rec = hdobj.get_data(kstpkper=(1, 50))
+    >>> rec = hdobj.get_data(kstpkper=(0, 50))
```
Comment (jdhughes-usgs) [inline flopy/utils/binaryfile.py:614]: (1,50) -> (0,49)?

**Review by jdhughes-usgs — COMMENTED** [review]
I am ok with this change although it may cause people who were using .kstpkper directly instead of using get_kstpkper().

**Comment by jdhughes-usgs** [comment]
> > maybe it would be safest just to clearly document that `kstpkper` is 1-based and leave it that way
> 
> I think it's preferable to leave `.kstpkper` as one-based in case someone is using it now and relying on that behavior. We might also consider deprecating it, as a way to continue removing remnant one-based information. If we deprecate/remove it, those using it will have to switch to the zero-based `get_kstpkper()` method.

We could convert the internal variable to `._kstpkper` and make `.kstpkper` a property that returns one-based values and issue a deprecation warning that `.kstpkper` will be removed.

File: flopy/utils/formattedfile.py (hunk 1 above)
Comment (wpbonelli) [reply flopy/utils/formattedfile.py]: updated

File: flopy/utils/binaryfile.py (hunk 2 above)
Comment (wpbonelli) [reply flopy/utils/binaryfile.py]: updated

**Comment by wpbonelli** [comment]
Pending more discussion and a decision on whether to deprecate I restored `kstpkper` to 1-based. This PR now just clarifies 0 vs 1-based indexing in docstrings and some other minor cleanup, no API changes or deprecations

**Pull request merged**